align/schema/checker.py

```diff
@@ -162,16 +163,20 @@ def bbox_vars(self, name):
         # width / height cannot be 0
         self.append(b.llx < b.urx)
         self.append(b.lly < b.ury)
-        # Do not overlap with other bboxes
-        for b2 in self._bbox_cache.values():
-            self.append(
-                self.Or(
-                    b.urx <= b2.llx,
-                    b2.urx <= b.llx,
-                    b.ury <= b2.lly,
-                    b2.ury <= b.lly,
-                )
-            )
+        if is_subcircuit:
+            self._bbox_subcircuit[name] = True
```

Comment:
Maybe I am missing something but looks like the constraint is only being checked once the subcircuit is instantiated. Not when the placement for the subcircuit is generated. Is that what we want?

Reply:
Can you check what happens if we do something silly like:
```json
        {
            "ratio_low": 0.1,
            "ratio_high": 0.5,
            "weight": 1,
            "const_name": "Aspect_Ratio"
        },
        {
            "ratio_low": 0.6,
            "ratio_high": 1.0,
            "weight": 1,
            "const_name": "Aspect_Ratio"
        }
]
```
I would expect the checker to throw an error as soon  as the second constraint is added.

Reply:
To clarify, I am not seeing any relationship between the subcircuit bbox and the instance bboxes within it.

Reply:
This constraint is not on an instance but on the subcircuit definition. Instantiation is not required as this variables are set in align/pnr/checker.py line 13 for each module (aka. subcircuit).

Reply:
Due to signal or power routing/grid, the subcircuit bbox could be larger than all the instances within it. Therefore, there is no explicit relationship in between. This constraints is checking the final bbox for the subcircuit.

Reply:
> Due to signal or power routing/grid, the subcircuit bbox could be larger than all the instances within it. Therefore, there is no explicit relationship in between. This constraints is checking the final bbox for the subcircuit.

We are basically relying on the placer/router engines to do the bbox computation correctly. If the placer computes a smaller bbox than the placed instances allow, we won't detect it. Conversely we can't write a placer using these constraints since the mathematical formulation is incomplete.

However, the chances of the first bug are extremely low (I hope 😄). And we haven't gotten to the point where we are doing SMT based placement yet. We can merge this in once the PnR issues are fixed.

Reply:
> Due to signal or power routing/grid, the subcircuit bbox could be larger than all the instances within it.

This can be easily captured by the way if we have access to the list of all instances within the subcircuit. I am working on this but it is proving to be a bit of work as it requires capturing parent relationships in our schema.

Reply:
> > Due to signal or power routing/grid, the subcircuit bbox could be larger than all the instances within it. Therefore, there is no explicit relationship in between. This constraints is checking the final bbox for the subcircuit.
> 
> We are basically relying on the placer/router engines to do the bbox computation correctly. If the placer computes a smaller bbox than the placed instances allow, we won't detect it. Conversely we can't write a placer using these constraints since the mathematical formulation is incomplete.
> 
> However, the chances of the first bug are extremely low (I hope 😄). And we haven't gotten to the point where we are doing SMT based placement yet. We can merge this in once the PnR issues are fixed.

Good point, Placer could make a mistake in calculating the bbox. Yet that is beyond the scope of this constraint 😄  Can we mark this as resolved?

Reply:
Yep!